Billing page followup

hackerkid opened the pull request.

rishig commented on templates/zilencer/upgrade.html:
```diff
@@ -35,9 +35,9 @@ <h3>{{ _("Payment schedule") }}</h3>
+                                    ${{ annual_plan_monthly_equalent_amount }}/user/month
                                     <div class="schedule-amount-2">
-                                        ($80/user/year)
+                                        (${{ annual_plan_amount }}/user/year)
```
Let's call these `cloud_monthly_price`, `cloud_annual_price`, `cloud_annual_price_per_month`

rishig commented on zilencer/models.py:
```diff
@@ -47,6 +47,10 @@ class Plan(models.Model):
     # The two possible values for nickname
     CLOUD_MONTHLY = 'monthly'
     CLOUD_ANNUAL = 'annual'
+    # Subscription amount in U.S Dollars
```
Let's hold off on this change; I'm not sure exactly how we'll want to do it, and we can hold off until we need it.

rishig commented on templates/zilencer/upgrade.html:
```diff
@@ -43,14 +43,14 @@ <h3>{{ _("Payment schedule") }}</h3>
                         </label>
                     </div>
-                    <p>You&rsquo;ll initially be charged <b><span id="charged_amount">XXX</span></b> for <b>{{ seat_count }}</b> users. You&rsquo;ll receive prorated charges and credits as users are added, deactivated, or become inactive.</p>
+                    <p>You&rsquo;ll initially be charged <b><span id="charged_amount">{{ annual_plan_amount * seat_count }}</span>$</b> for <b>{{ seat_count }}</b> users. You&rsquo;ll receive prorated charges and credits as users are added, deactivated, or become inactive.</p>
```
ah wow, I didn't realize you could multiply inside of the `{{  }}`!
A minor thing is the `$` should go at the beginning. (Possibly it should be written as `$50 USD`, i.e. with the USD?)

rishig commented on zilencer/lib/stripe.py:
```diff
@@ -122,7 +122,7 @@ def do_create_customer_with_payment_source(user: UserProfile, stripe_token: str)
     stripe_customer = stripe.Customer.create(
         description="%s (%s)" % (realm.string_id, realm.name),
-        metadata={'realm_id': realm.id, 'realm_str': realm.string_id},
+        metadata={'realm_id': realm.id, 'realm_str': realm.string_id, 'email': user.email},
```
This should be an argument to create: https://stripe.com/docs/api#customer_object-email

hackerkid commented:
@rishig These commits are ready for review.

rishig commented:
Thanks @hackerkid! I merged the first six commits, and pushed an updated version of the 7th. 

I think it would be useful to look at d32051b, compare it closely to your original version, and check that for each thing that is different you understand why I did it differently. For anything where you aren't sure just PM me and we can chat about it.

The updated version of the last commit doesn't pass tests, but can't be merged until 
https://chat.zulip.org/#narrow/stream/49-development-help/topic/realm.20audit.20log.20changes
is resolved anyway.

rishig commented:
Thanks @hackerkid! I spent a bit of time working on this, though didn't quite finish so pushing the progress here. The first commit I split out into #10332. 

The second commit is probably done, though needs a different commit message, like 
`billing: Make CTA buttons on /plans reflect current plan type.`

For the third commit we should compute the condition in home.py (the same place we compute show_invites). Let's show "Plans and pricing" to everyone if they are on LIMITED, show "Billing" to admins and the billing admin if they are on PREMIUM, and nothing in every other case. 

For the fourth commit, I unfortunately am not seeing the column on /activity, though maybe my changes messed it up.

hackerkid commented:
The current logged out Plans page shows `buy premium`. Should we change that as the sign up flow is just the same as clicking on "Current Plan"

![screenshot from 2018-08-21 16-21-09](https://user-images.githubusercontent.com/7190633/44397604-929be800-a55e-11e8-8275-fe7e7eb9266e.png)

I pushed some changes 
* Removed `billing_enabled` from template context.
* Updated the navbar commit to reflect your feedback as well as added tests.

> For the fourth commit, I unfortunately am not seeing the column on /activity, though maybe my changes messed it up.

It's showing up for me. Weird. 
![realm_activity](https://user-images.githubusercontent.com/7190633/44406575-9a1cba80-a579-11e8-90e5-9be8c83d1adc.png)

rishig commented:
Thanks @hackerkid! Merged the last commit; it's working for me now as well.
For the other two:
* Good point about the "Buy Premium". I wrote a note in the billing doc on how we can handle this. 
* For the nav bar change: I didn't think about the fact that we'd have to import from zilencer. Let's instead add a field is_billing_admin to UserProfile, default False, and set it in `do_create_customer_with_payment_source` if the user is not an admin.
We can also add a commit to remove Customer.billing_user.

hackerkid commented:
@rishig I made a minor change to the plans page to point `Buy premium` to `/billing` instead of `/new`  so that they would be redirected to login page. This works fine when a user is logged out and is accessing plans plage from their realm subdomain(eg https://hackinthenorth.zulipchat.com/stats). But when user tries to do so from root domain (https://zulipchat.com/stats) they would be redircted to find accounts.

Slack does this in two different ways.

* When I am logged into multiple organizations and I access slack.com/plans it redirects me to the plans page of the last organization(not sure thats how they decide the organization though) I visited. We have to write extra code to accomplish this.


* When I am not logged in and I access slack.com/plans it asks me to signin to an organization and redirects to plans page after I do that. In the current PR instead of asking people to sign in we redirect them to find accounts page. (In localhost people are redirected to locaclhost:9991 where all the accounts are listed). We need to create an extra page that asks people to sign to their organization instead of doing the find-accounts thing. I guess that is true for other cases where user have to sign in as well.

![screenshot from 2018-08-22 18-08-09](https://user-images.githubusercontent.com/7190633/44463771-66539a80-a636-11e8-94fb-4f0f42431496.png)

Is this how you like to see this implemented?

rishig commented on zerver/lib/actions.py:
```diff
@@ -3021,11 +3021,17 @@ def do_change_icon_source(realm: Realm, icon_source: str, log: bool=True) -> Non
-    RealmAuditLog.objects.create(event_type=RealmAuditLog.REALM_PLAN_TYPE_CHANGED,
-                                 realm=realm, acting_user=user, event_time=timezone_now(),
-                                 extra_data={'old_value': old_value, 'new_value': plan_type})
+    with transaction.atomic():
```
I'm not sure about this transaction.atomic; we don't seem to use them anywhere else in actions.py. Maybe worth opening a thread on #development-help.

If we do keep it, only the realm.save() and the RealmAuditLog entries need to be in it.

rishig commented on zerver/lib/actions.py:
```diff
@@ -3021,11 +3021,17 @@ def do_change_icon_source(realm: Realm, icon_source: str, log: bool=True) -> Non
+                                     extra_data={'old_value': old_value, 'new_value': plan_type})
+
+        if plan_type == Realm.PREMIUM:
+            # Set max_invites to a value that is impossible to run out in practical use
```
I think we don't need this comment, since it's clear from the code.

rishig commented on zerver/tests/test_realm.py:
```diff
@@ -341,6 +341,12 @@ def test_initial_plan_type(self) -> None:
         with self.settings(BILLING_ENABLED=False):
             self.assertEqual(Realm.SELF_HOSTED, do_create_realm('onpremise', 'onpremise').plan_type)
 
+    def test_initial_message_visibility_limit(self) -> None:
```
Let's combine these two tests into "test_initial_plan_type_settings"

hackerkid commented:
@rishig Updated

rishig commented:
Thanks @hackerkid! After playing around with this, let's add another button for /plans; a grayed out button like
![image](https://user-images.githubusercontent.com/890911/44631784-ed01c380-a925-11e8-9a7a-17ba74bfdd10.png)

Then we can do the following simpler behavior:
* On root domain: Current behavior
* On subdomain: Make it a logged-in-only page. I.e. should redirect to login, with `?next=...`

On subdomain, when logged in:
* plan_type self_hosted: Redirect to /plans/enterprise. For now it can be a page with just "Enterprise plans coming soon. Please contact support@zulipchat.com."
* plan_type limited: "Free" is a grey button, "Premium" is current behavior (green "Buy Premium" button that goes to /upgrade).
* plan_type premium or premium free: "Free" has nothing in the bottom section (no button or text), "Premium" is the grey button

So now the plans page no longer has to know anything about the user.

hackerkid commented:
@rishig Took care of the feedback.

rishig commented:
Thanks @hackerkid! Looks great. For next steps I'm going to wait until #10435 is merged (the /plans portion of this PR), which will likely happen when Tim is back in operation.

rishig commented:
fyi, merged part of this as #10435, and opened #10479 for the gear menu changes. Still reviewing /accounts/go.

rishig commented on zerver/forms.py:
```diff
@@ -349,3 +349,13 @@ def clean_emails(self) -> List[str]:
+
+class GoToRealmForm(forms.Form):
+    subdomain = forms.CharField(max_length=Realm.MAX_REALM_SUBDOMAIN_LENGTH, required=True)
+    next_param = forms.CharField(max_length=100, required=False)
```
Why do we use this instead of the `?next=` pattern we use for login?

rishig replied:
That does make sense, and seems like a good way of solving it. 
Do you understand what our login flow is doing then? It seems different from this.

rishig commented on zerver/forms.py:
```diff
@@ -349,3 +349,13 @@ def clean_emails(self) -> List[str]:
 
         return emails
+
+class GoToRealmForm(forms.Form):
```
Maybe `RealmRedirect` instead of `GoToRealm`?
A bit of a pain to change, but hopefully not too much since the names are mostly concentrated in one commit.

rishig commented:
Merged one more commit. Might be good to rebase the next time you look at this.

hackerkid commented:
@rishig Rebased and updated the `next` param backend to a style similiar to our dev_login.

rishig commented:
Great! Repushed, with minor changes. 
I think the next step is for Tim to review. Could you take these last two commits and open a new PR?
